tpls/desul/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,491 @@
+	|| std::is_same_v<T,unsigned int> || std::is_same_v<T,unsigned long long> \
+	|| std::is_same_v<T,float> || std::is_same_v<T,double>, R>;
+#else
+template <class T, class R>
+using acc_enable_if_supported_arithmetic_type = std::enable_if_t<std::is_same_v<T,int> \
+	|| std::is_same_v<T,unsigned int> || std::is_same_v<T,unsigned long long> \
+	|| std::is_same_v<T,float>, R>;
+#endif
+#else
+template <class T, class R>
+using acc_enable_if_supported_arithmetic_type = std::enable_if_t<std::is_arithmetic<T>::value, R>;
+#endif
+
+namespace desul {
+namespace Impl {
+
+// clang-format off
+//<editor-fold desc="device_atomic_fetch_{add,sub,mul,div,lshift,rshift,mod,max,min,and,or,xor}">
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_add(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_inc(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_sub(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_dec(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_mul(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr *= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_div(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr /= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_lshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr << val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_rshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr >> val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_fetch_mod(
```

Comment:
Do I understand correctly that there is no path forward with the current acc specification to support atomic modulo, min, nor max?

Reply:
The current OpenACC standard supports atomic min and max operations only for Fortran, and atomic modulo is not supported at all.

Reply:
Even though OpenACC C/C++ standard does not support atomic min/max/mod operations, the Kokkos OpenACC backend supports them using either CUDA intrinsics (for NVHPC implementation) or generic lock-free fetch-op implementation in Lock_Free_Fetch_Op.hpp.